### PR TITLE
Deprecate the DependencyFactory class for removal in Gradle 8.0.

### LINKED ISSUE #21886: Deprecate `org.gradle.api.internal.artifacts.dsl.dependencies.DependencyFactory`
This class will be removed in Gradle 8.0, see: #21880

### PR DESCRIPTION
**Pull request by tresat**
Fixes #21886

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/core/src/main/java/org/gradle/api/internal/DependencyClassPathProvider.java [hunk 1]
```diff
@@ -53,17 +49,18 @@ public DependencyClassPathProvider(ModuleRegistry moduleRegistry, PluginModuleRe
     }
 
     @Override
+    @SuppressWarnings("deprecation")
```
Comment (octylFractal) [inline subprojects/core/src/main/java/org/gradle/api/internal/DependencyClassPathProvider.java:52]: Why does this have a deprecation warning?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/notations/DependencyNotationParser.java [hunk 2]
```diff
@@ -79,24 +77,6 @@ public static DependencyNotationParser create(Instantiator instantiator,
             .converter(mapNotationConverter)
             .fromType(FileCollection.class, filesNotationConverter)
             .fromType(Project.class, projectNotationConverter)
-            .fromType(DependencyFactory.ClassPathNotation.class, new NotationConverter<DependencyFactory.ClassPathNotation, Dependency>() {
```
Comment (octylFractal) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/notations/DependencyNotationParser.java:82]: I am not sure why this was removed. This is still needed to support the notation, right?

**Review by octylFractal — CHANGES_REQUESTED** [review]
2 parts of this seem a little odd, plus the changes to remove the static imports seem a bit extraneous.

File: subprojects/core/src/main/java/org/gradle/api/internal/DependencyClassPathProvider.java (hunk 1 above)
Comment (tresat) [reply subprojects/core/src/main/java/org/gradle/api/internal/DependencyClassPathProvider.java:52]: It's from before I merged the `DependencyFactoryInternal`, I must have missed deleting it during the merge.  It's useless now, I'll remove it.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/notations/DependencyNotationParser.java (hunk 2 above)
Comment (tresat) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/notations/DependencyNotationParser.java:82]: Isn't line 100 enough?

**Comment by tresat** [comment]
> the changes to remove the static imports seem a bit extraneous.

I needed to use the FQN instead of an import so I could suppress deprecation warnings, back when there was no `DependencyFactoryInternal` and I was referencing the deprecated `DependencyFactory`.

They do look extraneous now, I'll revert them.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/notations/DependencyNotationParser.java (hunk 2 above)
Comment (octylFractal) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/notations/DependencyNotationParser.java:82]: That's the new notation. Won't using the deprecated notation fail without this converter?

Comment (tresat) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/notations/DependencyNotationParser.java:82]: I thought that was the idea, it's marked for removal in Gradle 8 below.

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@octylFractal - Is `DependencyClassPathNotationConverter#create(DependencyFactoryInternal.ClassPathNotation)` missing a case for `LOCAL_GROOVY`?  

The Gretty 4.x smoke test is failing locally for me with:

```
Execution failed for task ':appBeforeIntegrationTest'.
> Cannot convert the provided notation to an object of type Dependency: LOCAL_GROOVY.
  The following types/formats are supported:
    - String or CharSequence values, for example 'org.gradle:gradle-core:1.0'.
    - Maps, for example [group: 'org.gradle', name: 'gradle-core', version: '1.0'].
    - FileCollections, for example files('some.jar', 'someOther.jar').
    - Projects, for example project(':some:project:path').
    - ClassPathNotation, for example gradleApi().
  
  Comprehensive documentation on dependency notations is available in DSL reference for DependencyHandler type - if you are so inclined.
```

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle cancel

**Comment by tresat** [comment]
@bot-gradle test this

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GrettySmokeTest.groovy [hunk 3]
```diff
@@ -60,7 +61,8 @@ class GrettySmokeTest extends AbstractPluginValidatingSmokeTest {
 
         when:
         def result = runner('checkContainerUp')
-            .expectDeprecationWarning(
+            .expectDeprecationWarningIf(
```
Comment (octylFractal) [inline subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GrettySmokeTest.groovy:64]: This warning isn't present anymore, it should be a hard error instead if it happens in any of the tests.

**Review by octylFractal — CHANGES_REQUESTED** [review]

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GrettySmokeTest.groovy (hunk 3 above)
Comment (tresat) [reply subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GrettySmokeTest.groovy:64]: That makes sense, we'll never go back to testing an earlier version.

**Comment by tresat** [comment]
@bot-gradle test and merge

**Comment by tresat** [comment]
@bot-gradle test and merge

**Pull request merged**